File: cli/manifest.py
#!/usr/bin/env python3
"""Orchestrator-owned manifest helpers (stdlib-only).

Manifest tracks workflow state per MASTER_SPEC 11.1. All writes are
atomic (tmp file + os.replace) so kill -9 cannot leave a half-written
manifest (stepper + resume live in `cli/run.py` since PR7).
"""

import hashlib
import json
import os
import tempfile
from datetime import datetime, timezone
from pathlib import Path

MANIFEST_NAME = "manifest.json"
WORKFLOW_VERSION = "1.0.0"

GREENFIELD_WORKFLOW = "greenfield-full-lite"
BROWNFIELD_WORKFLOW = "brownfield-fix"


def utcnow():
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def default_workflow(green=True):
    return GREENFIELD_WORKFLOW if green else BROWNFIELD_WORKFLOW


def default_budgets():
    return {
        "tokens": {"limit": 800000, "used": 0},
        "spendUSD": {"limit": 25.0, "used": 0.0},
    }


def genesis(core_version, workflow, budgets=None):
    """Build a genesis manifest dict (not yet saved)."""
    return {
        "coreVersion": core_version,
        "workflow": workflow,
        "workflowVersion": WORKFLOW_VERSION,
        "artifacts": {},
        "gates": {},
        "budgets": budgets if budgets is not None else default_budgets(),
        "retries": {},
        "checkpoints": [],
        "initializedAt": utcnow(),
    }


def manifest_path(project_dir):
    return Path(project_dir) / ".shiploom" / MANIFEST_NAME


def load(project_dir):
    """Load manifest or raise (FileNotFoundError / ValueError on bad JSON)."""
    path = manifest_path(project_dir)
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except FileNotFoundError:
        raise
    except (OSError, ValueError) as exc:
        raise ValueError("unreadable manifest %s: %s" % (path, exc))


def save(project_dir, data):
    """Atomically write the manifest (kill-safe)."""
    path = manifest_path(project_dir)
    path.parent.mkdir(parents=True, exist_ok=True)
    fd, tmp = tempfile.mkstemp(dir=str(path.parent), prefix=".manifest-", suffix=".tmp")
    try:
        with os.fdopen(fd, "w", encoding="utf-8") as fh:
            json.dump(data, fh, indent=2, sort_keys=True)
            fh.write("\n")
        os.replace(tmp, path)
    except BaseException:
        try:
            os.unlink(tmp)
        except OSError:
            pass
        raise
    return path


def sha256_file(path):
    digest = hashlib.sha256()
    with open(path, "rb") as fh:
        for chunk in iter(lambda: fh.read(65536), b""):
            digest.update(chunk)
    return "sha256:" + digest.hexdigest()

File: internal/manifest/manifest.go
// Package manifest ports cli/manifest.py (stdlib only).
package manifest

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"time"

	"github.com/shiploom/ai-builder/internal/jsoncanon"
)

const (
	// ManifestName is the orchestrator-owned state file.
	ManifestName = "manifest.json"
	// WorkflowVersion stamps genesis manifests.
	WorkflowVersion = "1.0.0"
	// GreenfieldWorkflow is the default greenfield flow.
	GreenfieldWorkflow = "greenfield-full-lite"
	// BrownfieldWorkflow is the default brownfield flow.
	BrownfieldWorkflow = "brownfield-fix"
)

// Utcnow mirrors Python utcnow(): "2026-09-17T10:00:00Z".
func Utcnow() string {
	return time.Now().UTC().Format("2006-01-02T15:04:05Z")
}

// DefaultWorkflow mirrors default_workflow().
func DefaultWorkflow(green bool) string {
	if green {
		return GreenfieldWorkflow
	}
	return BrownfieldWorkflow
}

// DefaultBudgets mirrors default_budgets(). Floats use literals so the
// canonical writer emits "25.0"/"0.0" exactly like CPython.
func DefaultBudgets() map[string]any {
	return map[string]any{
		"tokens":   map[string]any{"limit": int64(800000), "used": int64(0)},
		"spendUSD": map[string]any{"limit": json.Number("25.0"), "used": json.Number("0.0")},
	}
}

// Genesis builds a genesis manifest dict (not yet saved).
func Genesis(coreVersion, workflow string, budgets map[string]any) map[string]any {
	if budgets == nil {
		budgets = DefaultBudgets()
	}
	return map[string]any{
		"coreVersion":     coreVersion,
		"workflow":        workflow,
		"workflowVersion": WorkflowVersion,
		"artifacts":       map[string]any{},
		"gates":           map[string]any{},
		"budgets":         budgets,
		"retries":         map[string]any{},
		"checkpoints":     []any{},
		"initializedAt":   Utcnow(),
	}
}

// ManifestPath mirrors manifest_path().
func ManifestPath(projectDir string) string {
	return filepath.Join(projectDir, ".shiploom", ManifestName)
}

// Load mirrors load(): os.ErrNotExist when missing, wrapped message when corrupt.
func Load(projectDir string) (map[string]any, error) {
	path := ManifestPath(projectDir)
	raw, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, err
		}
		return nil, fmt.Errorf("unreadable manifest %s: %s", path, fsErrText(err))
	}
	doc, err := jsoncanon.Decode(raw)
	if err != nil {
		// Mirror "unreadable manifest %s: %s"; suffix is engine-specific.
		return nil, fmt.Errorf("unreadable manifest %s: %s", path, err)
	}
	obj, ok := doc.(map[string]any)
	if !ok {
		return nil, fmt.Errorf("unreadable manifest %s: not an object", path)
	}
	return obj, nil
}

// Save atomically writes the manifest (kill-safe): temp file + rename,
// best-effort temp cleanup on failure.
func Save(projectDir string, data map[string]any) (string, error) {
	path := ManifestPath(projectDir)
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return "", err
	}
	out, err := jsoncanon.Marshal(data)
	if err != nil {
		return "", err
	}
	out = append(out, '\n')
	tmp, err := os.CreateTemp(filepath.Dir(path), ".manifest-*.tmp")
	if err != nil {
		return "", err
	}
	tmpName := tmp.Name()
	if _, err := tmp.Write(out); err != nil {
		tmp.Close()
		os.Remove(tmpName)
		return "", err
	}
	if err := tmp.Close(); err != nil {
		os.Remove(tmpName)
		return "", err
	}
	// os.Rename cannot replace an existing file on Windows, so remove
	// first (POSIX keeps atomic replace; the crash window on Windows is
	// accepted — manifests are rewritten on every run).
	if err := os.Remove(path); err != nil && !os.IsNotExist(err) {
		os.Remove(tmpName)
		return "", err
	}
	if err := os.Rename(tmpName, path); err != nil {
		os.Remove(tmpName)
		return "", err
	}
	return path, nil
}

// Sha256File mirrors sha256_file().
func Sha256File(path string) (string, error) {
	fh, err := os.Open(path)
	if err != nil {
		return "", err
	}
	defer fh.Close()
	digest := sha256.New()
	if _, err := io.Copy(digest, fh); err != nil {
		return "", err
	}
	return "sha256:" + hex.EncodeToString(digest.Sum(nil)), nil
}

func fsErrText(err error) string {
	if pe, ok := err.(*os.PathError); ok {
		return pe.Err.Error()
	}
	return err.Error()
}
